pkg/operator/ceph/cluster/monitoring.go

```diff
@@ -0,0 +1,167 @@
+			logger.Infof("enabling ceph mon monitoring goroutine for cluster %q", cluster.Namespace)
+			go healthChecker.Check(cluster.monitoringChannels["mon"].stopChan)
+
+			// Set the flag that it is already running
+			cluster.monitoringChannels["mon"].monitoringRunning = true
+		}
+	}
+}
+
+func (c *ClusterController) startCephOSDMonitoring(cluster *cluster, cephUser string) {
+
+	if !cluster.Spec.External.Enable {
+		if osdHealth, ok := cluster.monitoringChannels["osd"]; ok {
+			if osdHealth.monitoringRunning {
+				// If the goroutine was running but the CR was updated to stop the monitoring we need to close the channel
+				if cluster.Spec.HealthCheck.DaemonHealth.ObjectStorageDaemon.Disabled {
+					// close the channel so the goroutine can stop
+					close(cluster.monitoringChannels["osd"].stopChan)
+					// Set monitoring to false since it's not running anymore
+					cluster.monitoringChannels["osd"].monitoringRunning = false
+				} else {
+					logger.Infof("ceph osd health go routine is already running for cluster %q", cluster.Namespace)
+				}
+			} else {
+				// if not already running and not disabled, we run it
+				if !cluster.Spec.HealthCheck.DaemonHealth.ObjectStorageDaemon.Disabled {
+					c.osdChecker = osd.NewOSDHealthMonitor(c.context, cluster.Namespace, cluster.Spec.RemoveOSDsIfOutAndSafeToRemove, cluster.Spec.HealthCheck)
+					logger.Infof("enabling ceph osd monitoring goroutine for cluster %q", cluster.Namespace)
+					go c.osdChecker.Start(cluster.monitoringChannels["osd"].stopChan)
+
+					// Set the flag that it is already running
+					cluster.monitoringChannels["osd"].monitoringRunning = true
+				}
+			}
+		} else {
+			// If the mon does not exist in the map, this is a first deployment or an operator restart
+			// So we check the desired state from the CR and run it if necessary
+			//
+			// If the mon monitoring is enabled
+			if !cluster.Spec.HealthCheck.DaemonHealth.ObjectStorageDaemon.Disabled {
+				cluster.monitoringChannels["osd"] = &clusterHealth{
+					stopChan:          make(chan struct{}),
+					monitoringRunning: false,
+				}
+
+				c.osdChecker = osd.NewOSDHealthMonitor(c.context, cluster.Namespace, cluster.Spec.RemoveOSDsIfOutAndSafeToRemove, cluster.Spec.HealthCheck)
+				logger.Infof("enabling ceph osd monitoring goroutine for cluster %q", cluster.Namespace)
+				go c.osdChecker.Start(cluster.monitoringChannels["osd"].stopChan)
+
+				// Set the flag that it is already running
+				cluster.monitoringChannels["osd"].monitoringRunning = true
+			}
+		}
+	}
+
+}
+
+func (c *ClusterController) startCephStatusMonitoring(cluster *cluster, cephUser string) {
+	// Start the mon health checker
+	if monHealth, ok := cluster.monitoringChannels["status"]; ok {
+		if monHealth.monitoringRunning {
+			// If the goroutine was running but the CR was updated to stop the monitoring we need to close the channel
+			if cluster.Spec.HealthCheck.DaemonHealth.Monitor.Disabled {
```

Comment:
Seems like some of this could be factored into common methods between the different status checkers. For example, lines 129-136 look exactly the same as in the other methods except the name of a couple fields.

Reply:
Potentially but not so much, I've tried to factor a bit but since functions don't have the same signature, it's complex, using an interface can help but then we need to call each method in a goroutine individually. Not sure it's worth the effort.

Reply:
Actually, I have something!